xla/service/gpu/cudnn_norm_rewriter.cc

```diff
@@ -133,6 +209,162 @@ bool CalculatesExpectation(const HloInstruction* instr) {
          ((actual_r_nelems + r_nelems) * numerical_epsilon);
 }
 
+// Recursively traverses the graph across converts, bitcasts and reshapes
+// starting from instr, and returns the operand of custom_call with index
+// operand_idx. Returns nullptr if the prescribed operand of custom_call cannot
+// be reached from instr.
+HloInstruction* FindOperandRecursive(HloInstruction* instr,
+                                     HloInstruction* custom_call,
+                                     absl::flat_hash_set<int>& visited_instrs,
+                                     int operand_idx,
+                                     HloInstruction* transpose) {
+  visited_instrs.emplace(instr->unique_id());
+  const absl::flat_hash_set<HloOpcode> supported_ops = {
+      HloOpcode::kConvert, HloOpcode::kBitcast, HloOpcode::kReshape};
+  // Look for the Custom Call among the users of instr.
+  for (HloInstruction* user : instr->users()) {
+    if (user == custom_call && user->operand(operand_idx) == instr) {
+      return instr;
+    }
+    if ((supported_ops.contains(user->opcode()) || user == transpose) &&
+        !visited_instrs.contains(user->unique_id())) {
+      return FindOperandRecursive(user, custom_call, visited_instrs,
+                                  operand_idx, transpose);
+    }
+  }
+  // Ascend the graph if the Custom Call is not found and instr is a
+  // conversion, bitcast or reshape.
+  if (supported_ops.contains(instr->opcode())) {
+    return FindOperandRecursive(instr->mutable_operand(0), custom_call,
+                                visited_instrs, operand_idx, transpose);
+  }
+  return nullptr;
+}
+HloInstruction* FindOperandRecursive(HloInstruction* instr,
+                                     HloInstruction* custom_call,
+                                     int operand_idx,
+                                     const NormMetaDataMap& norm_meta_data) {
+  absl::flat_hash_set<int> visited_instrs;
+  auto custom_call_meta_data = norm_meta_data.find(custom_call->unique_id());
+  if (custom_call_meta_data == norm_meta_data.end()) {
+    return nullptr;
+  }
+  return FindOperandRecursive(instr, custom_call, visited_instrs, operand_idx,
+                              custom_call_meta_data->second.x_transpose);
+}
+
+// Maps the dimensions of the reduction in the calculation of Dbias from the
+// potentially reshaped DY to the original DY, i.e. the input to the reduction,
+// assuming the transformation can be described by reshapes.
+std::vector<int64_t> MapReduceDims(HloInstruction* dy,
```

Comment:
This function is quite confusing as to what it does, including the comment. How about improving the comment and choosing a slightly better function name and parameter names?

What about something like this:
```
/**
 * Computes the dimensions over which `reshaped_shape` is reduced, if possible.
 * `original_shape` is reduced over dimensions `reduce_dims`.
 * `reshaped_shape` is a strict reshape of `original_shape`.
 *
 * @param original_shape  The original shape which is reduced over dimensions
 *                        `reduce_dims`.
 * @param reduce_dims     The dimensions `original_shape` is reduced over.
 * @param reshaped_shape  A reshape of `original_shape`.
 * 
 * @return                The dimensions over which `reshaped_shape` is reduced, if
 *                        possible. If not possible, returns std::nullopt.
 */
std::optional<std::vector<int64_t>> AxesOfReductionForReshapedShape(
    const Shape &original_shape, absl::Span<int64_t> reduce_dims,
    const Shape &reshaped_shape) {
  //
}
```

Also:
- The function should return an optional, because a mapping of reduction dimensions is not always possible.
- Pass in an `absl::Span`

Reply:
Ah yeah, also pass a `Shape` rather than an `HloInstruction`